topics/go/language/embedding/exercises/template1/template1.go

```diff
@@ -1,53 +1,88 @@
-// When implementing the find method for hockey, call into the find method
-// for the embedded sport type to check the embedded fields first. Then create
-// two hockey values inside the slice of haystacks and perform the search.
+// This program defines a type Feed with two methods: Count and Fetch. Create a
+// new type CachingFeed that embeds *Feed but overrides the Fetch method.
+//
+// The CachingFeed type should have a map of Documents to limit the number of
+// calls to Feed.Fetch.
 package main
 
-import "strings"
+import (
+	"fmt"
+	"log"
+	"time"
+)
 
-// finder defines the behavior required for performing matching.
-type finder interface {
-	find(needle string) bool
+// Document is the core data model we are working with.
+type Document struct {
+	Key   string
+	Title string
 }
 
-// sport represents a sports team.
-type sport struct {
-	team string
-	city string
+// ==================================================
+
+// Feed is a type that knows how to fetch Documents.
+type Feed struct{}
+
+// Count tells how many documents are in the feed.
+func (f *Feed) Count() int {
+	return 42
+}
+
+// Fetch simulates looking up the document specified by key. It is slow.
+func (f *Feed) Fetch(key string) (Document, error) {
+	time.Sleep(1000 * time.Millisecond)
+
+	doc := Document{
+		Key:   key,
+		Title: "Title for " + key,
+	}
+	return doc, nil
 }
 
-// find checks the value for the specified term.
-func (s *sport) find(needle string) bool {
-	return strings.Contains(s.team, needle) || strings.Contains(s.city, needle)
+// ==================================================
+
+// FetchCounter is the behavior we depend on for our process function.
+type FetchCounter interface {
+	Fetch(key string) (Document, error)
+	Count() int
 }
 
-// Declare a struct type named hockey that represents specific
-// hockey information:
-// - Have it embed the sport type first.
-// - Have it include a field with the country of the team.
+func process(fc FetchCounter) {
+	fmt.Printf("There are %d documents\n", fc.Count())
 
-// find checks the value for the specified term.
-func ( /* receiver type */ ) find(needle string) bool {
+	keys := []string{"a", "a", "a", "b", "b", "b"}
 
-	// Make sure you call into find method for the embedded sport type.
+	for _, key := range keys {
+		doc, err := fc.Fetch(key)
+		if err != nil {
+			log.Printf("Could not fetch %s : %v", key, err)
+			return
+		}
 
-	// Implement the search for the new fields.
-	return false
+		fmt.Printf("%s : %v\n", key, doc)
+	}
 }
 
+// ==================================================
+
+// Define a new type CachingFeed that embeds *Feed and has a map[string]Document.
+
+// Define a function NewCachingFeed which initializes a *CachingFeed for use.
+
+// Override the Fetch method with a new method that consults the map.
+
+// ==================================================
+
 func main() {
 
-	// Define the term to find.
+	fmt.Println("Using Feed directly")
+	var f Feed
```

Comment:
@jcbwlkr it bothers me the call to the NewCachingFeed is using the same Concrete feed value.

Reply:
@ardan-bkennedy why does this bother you? And what would you different? Use `Feed{}` literals?

Reply:
@ardan-bkennedy Be careful using words like "never", Bill. I literally don't see how this is a problem. 
- The `Feed` type implements its methods with a pointer receiver.
- The `CachingFeed` embeds a `*Feed`. It is meant as a decorator around an existing `*Feed` to override some behavior to provide some level of caching.
- This shows that embedding+overriding behavior very well in a way that is honestly intuitive.

Sure there are probably some types that exist where this wont work. I will open a new PR where this is changed to 

```go
func main() {
	fmt.Println("Using Feed directly")
	process(Feed{})

	fmt.Println("Using CachingFeed")
	c := NewCachingFeed(&Feed{})
	process(c)
}
```

Reply:
I appreciate the "Be careful using words like "never", Bill." comment but you were sharing the same concrete value to process in two different contexts, hence the word "never".

This is better since each calling is getting its own value.